mne/preprocessing/maxwell.py

```diff
@@ -77,38 +76,40 @@ def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
         raise RuntimeError('Maxwell filter cannot handle compensated '
                            'channels.')
 
-    all_coils, _, _, meg_info = _prep_meg_channels(raw.info, accurate=True,
+    # Get channels to use in multipolar moment calculation
+    good_chs = pick_types(raw.info, exclude='bads')
+    # Figure out channels to use in data reconstruction
+    recon_chs = pick_types(raw.info, exclude=[]) if reconstruct else good_chs
+
+    meg_coils, _, _, meg_info = _prep_meg_channels(raw.info, accurate=True,
                                                    elekta_defs=True)
 
-    # Create coil list and pick MEG channels
-    picks = [raw.info['ch_names'].index(coil['chname'])
-             for coil in all_coils]
-    coils = [all_coils[ci] for ci in picks]
     raw.preload_data()
-
-    data, _ = raw[picks, :]
+    data, _ = raw[good_chs, :]
 
     # Magnetometers (with coil_class == 1.0) must be scaled by 100 to improve
     # numerical stability as they have different scales than gradiometers
-    coil_scale = np.ones(len(picks))
-    coil_scale[np.array([coil['coil_class'] == 1.0 for coil in coils])] = 100.
+    coil_scale = np.ones(len(meg_coils))
+    coil_scale[np.array([coil['coil_class'] == 1.0
+                         for coil in meg_coils])] = 100.
 
     # Compute multipolar moment bases
     origin = np.array(origin) / 1000.  # Convert scale from mm to m
-    S_in, S_out = _sss_basis(origin, coils, int_order, ext_order)
+    S_in, S_out = _sss_basis(origin, meg_coils, int_order, ext_order)
     S_tot = np.c_[S_in, S_out]
 
     # Pseudo-inverse of total multipolar moment basis set (Part of Eq. 37)
     pS_tot = pinv(S_tot, cond=1e-15)
     # Compute multipolar moments of (magnetometer scaled) data (Eq. 37)
-    mm = np.dot(pS_tot, data * coil_scale[:, np.newaxis])
+    mm = np.dot(pS_tot[:, good_chs], data * coil_scale[good_chs, np.newaxis])
+
     # Reconstruct data from internal space (Eq. 38)
-    recon = np.dot(S_in, mm[:S_in.shape[1], :])
+    recon = np.dot(S_in[recon_chs, :], mm[:S_in.shape[1], :])
 
     # Return reconstructed raw file object
     raw_sss = _update_sss_info(raw.copy(), origin, int_order, ext_order,
                                data.shape[0])
-    raw_sss._data[picks, :] = recon / coil_scale[:, np.newaxis]
+    raw_sss._data[recon_chs, :] = recon / coil_scale[recon_chs, np.newaxis]
```

Comment:
I'm not 100% following. If we always reconstruct the bad channels, then why would we also clear them? Or do you mean, we should always clear _and then_ overwrite them with the reconstructed channel information?

Maybe it will be easier this way. If we have 306 channels and 5 are bad, do we want to return data with 
1) 306 channels (306 maxwell filtered and 301 used in MM calc; 5 left out of multipolar moment estimate but reconstructed from the MMs)
2) 306 channels (301 maxwell filtered and used in the MM calc, 5 left untouched)
3) 301 channels (301 maxwell filtered and used in MM calc, 5 entirely cleared out of `raw`)

As it stands, the code can do 1) and 2) depending on the `reconstruct` flag.